date_range/models/date_range.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2016 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class DateRange(models.Model):
+    _name = "date.range"
+    _description = "Date Range"
+    _order = "type_name,date_start"
+
+    @api.model
+    def _default_company(self):
+        return self.env['res.company']._company_default_get('date.range')
+
+    name = fields.Char(required=True, translate=True)
+    date_start = fields.Date(string='Start date', required=True)
+    date_end = fields.Date(string='End date', required=True)
+    type_id = fields.Many2one(
+        comodel_name='date.range.type', string='Type', index=1, required=True,
+        ondelete='restrict', domain="['|', ('company_id', '=', company_id), "
+                                    "('company_id', '=', False)]")
+    type_name = fields.Char(
+        related='type_id.name', store=True, string="Type Name")
+    company_id = fields.Many2one(
+        comodel_name='res.company', string='Company', index=1,
+        default=_default_company)
+    active = fields.Boolean(
+        help="The active field allows you to hide the date range without "
+        "removing it.", default=True)
+
+    _sql_constraints = [
+        ('date_range_uniq', 'unique (name,type_id, company_id)',
+         'A date range must be unique per company !')]
+
+    @api.onchange('company_id', 'type_id')
+    def _onchange_company_id(self):
```

Comment:
in odoo 13 you shoulr replace onchange method by computed fields with 'readonly=False'. It's more efficient...